File: python/day04/day04.py
import os
import time
import re


def main():
    # input
    print(os.getcwd())
    day = "04"
    year = "2022"
    input_file = f'../inputs/day{day}.txt'
    print(input_file)
    part1, part2 = 0, 0
    with open(input_file) as f:
        lines = f.read().splitlines()

    start_time = time.time()

    rgs = [list(map(int, re.split('-|,', l))) for l in lines]

    def contains(r): return (r[0] >= r[2] and r[1] <= r[3]) or (
        r[2] >= r[0] and r[3] <= r[1])

    def overlap(r): return not(r[1] < r[2] or r[3] < r[0])
    part1 = sum(map(contains, rgs))
    part2 = sum(map(overlap, rgs))

    duration = int((time.time() - start_time) * 1000000)
    header = "#" * 21
    print(
        f"{header}\n* AoC {year} - Day {day} *\n{header}\n\nPart 1:\t{part1}\nPart 2:\t{part2}\nTime:\t{duration // 1000} ms")


if __name__ == "__main__":
    main()

File: js/day04/day04.js
const { input } = require("./input");
const day = 4;
const year = 2022;
let part1 = 0;
let part2 = 0;


let parseInput = (arr) => {
  return arr
    .trim()
    .split(",")
    .map((s) => s.split(`-`).map((i) => parseInt(i, 10)))
    .flat();
};

let containEachOther = ([x1,y1,x2,y2]) => {
  return (x1 >= x2 && y1 <= y2) || (x2 >= x1 && y2 <= y1);
};

// in order to check if two ranges overlap, it is easier to test for NOT overlapping.
// looking at the data, we can safely assume x1<=y1 && x2 << y2
let overlap = ([x1,y1,x2,y2]) => {
    return !(y2<x1 || y1<x2);
}

part1 = input
  .map((l) => parseInput(l))
  .filter((l) => containEachOther(l)).length;

part2 = input
  .map((l) => parseInput(l))
  .filter((l) => overlap(l)).length;

console.log(
  `Advent of Code ${year} - Day ${String(day).padStart(2, "0")} \n${"#".repeat(
    28
  )}\nPart 1:\t${part1}\nPart 2:\t${part2}`
);
